### PR TITLE
Feature/grippe

### PR DESCRIPTION
**Pull request by AlonSchwierz**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/gripper/Commands/toggleSolenoid.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package frc.robot.subsystems.gripper.Commands;
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.subsystems.gripper.Gripper;
+
+public class toggleSolenoid extends CommandBase {
+    private Gripper gripper;
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/gripper/Commands/toggleSolenoid.java]: Please initialize the gripper in the constructor and `addRequirements`

File: src/main/java/frc/robot/subsystems/gripper/Commands/toggleSolenoid.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package frc.robot.subsystems.gripper.Commands;
+
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.subsystems.gripper.Gripper;
+
+public class toggleSolenoid extends CommandBase {
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/gripper/Commands/toggleSolenoid.java]: Should be instant command

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import edu.wpi.first.wpilibj.PneumaticsModuleType;
+import edu.wpi.first.wpilibj.Solenoid;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Ports;
+import frc.robot.subsystems.LoggedSubsystem;
+
+public class Gripper extends LoggedSubsystem<GripperLoggedInputs> {
+    private GripperLoggedInputs gripperLoggedInputs;
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/gripper/Gripper.java]: There is no need for this, you have thus from the super class

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    private final Solenoid leftSolenoid = new Solenoid(PneumaticsModuleType.CTREPCM, Ports.Gripper.LEFT_SOLENOID);
+    private final Solenoid rightSolenoid = new Solenoid(PneumaticsModuleType.CTREPCM, Ports.Gripper.RIGHT_SOLENOID);
+
+    private final NetworkTable table = NetworkTableInstance.getDefault().getTable("gripper");
```
Comment (barel13) [inline src/main/java/frc/robot/subsystems/gripper/Gripper.java]: Well... you don't use this class

**Review by barel13 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Ports.java [hunk 5]
```diff
@@ -1,5 +1,8 @@
 package frc.robot;
 
 public final class Ports {
-
+public class Gripper{
+    public static final int LEFT_SOLENOID = 0;
+    public static final int RIGHT_SOLENOID = 0;
+}
```
Comment (katzuv) [inline src/main/java/frc/robot/Ports.java:7]: These should be in a ports file in the gripper package.

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * return the current state of the solenoids
+     */
+    public void getRightSolenoid() {
+        rightSolenoid.get();
+    }
+
+    public void getLeftSolenoid() {
+        leftSolenoid.get();
+    }
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/gripper/Gripper.java:61]: What is a use case for these methods?

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * changes the current state to the other state
+     */
+
+    public void toggleSolenoid() {
+        leftSolenoid.toggle();
+        rightSolenoid.toggle();
+
+    }
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/gripper/Gripper.java]: Implementation details should stay hidden. That means that end-users don't need to know that the gripper is using solenoids.
So, these methods should be called something similar to "open"/"close" and "toggle".

File: src/main/java/frc/robot/subsystems/gripper/Commands/toggleSolenoid.java (hunk 2 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/gripper/Commands/toggleSolenoid.java]: The whole class can be replace with an inline command: `new InstantCommand(gripper::toggleSolenoid)`.

**Review by katzuv — CHANGES_REQUESTED** [review]
Nice! A few notes:
- "grippe" in the branch name should be fixed to "gripper". Do this through GitHub.
- The PR's title should be human-readable, not the same as the branch name.
- The `Gripper` class lacks a `getInstance()` method.
- I'd add a "close" method as well to the `Gripper` class.

**Comment by katzuv** [comment]
Also, note that the assignees' list contains the people who _work_ on the PR -- this is you (@AlonSchwierz). If you want a review from someone, just add them to the reviewers' list 😊.

File: src/main/java/frc/robot/Ports.java (hunk 5 above)
Comment (GaiaZano05) [reply src/main/java/frc/robot/Ports.java:7]: We decided to have a shared Port file and a seperate Constants file.

File: src/main/java/frc/robot/subsystems/gripper/Commands/toggleSolenoid.java (hunk 1 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/subsystems/gripper/Commands/toggleSolenoid.java]: done.

File: src/main/java/frc/robot/subsystems/gripper/Commands/toggleSolenoid.java (hunk 2 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/subsystems/gripper/Commands/toggleSolenoid.java]: done.

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java (hunk 3 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/subsystems/gripper/Gripper.java]: done.

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java (hunk 4 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/subsystems/gripper/Gripper.java]: OK.

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java (hunk 6 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/subsystems/gripper/Gripper.java:61]: Well, the only case I can think of is when the arm engages in order to pick up cones, if we want to do that we should know whether the gripper is closed or open.

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java (hunk 7 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/subsystems/gripper/Gripper.java]: done

File: src/main/java/frc/robot/subsystems/gripper/Gripper.java (hunk 6 above)
Comment (vichik123) [reply src/main/java/frc/robot/subsystems/gripper/Gripper.java:61]: You get a parameter `GripperLogInputs` in the constructor, and pass it to the super constructor. However, the logger inputs should be instantiated right in the constructor, without being passed as a parameter.

File: src/main/java/frc/robot/subsystems/gripper/GripperLoggedInputs.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package frc.robot.subsystems.gripper;
+
+import org.littletonrobotics.junction.LogTable;
+import org.littletonrobotics.junction.inputs.LoggableInputs;
+
+public class GripperLoggedInputs implements LoggableInputs {
+    boolean leftSolenoid;
+    boolean rightSolenoid;
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/gripper/GripperLoggedInputs.java:8]: Maybe you should give more descriptive names for these fields, like - `leftSolenoidExtended`.

**Review by vichik123 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**